proposals/extensions.md

```diff
@@ -399,54 +399,194 @@ These requirements need more refinement as implementation progresses, and may ne
+- The signatures of the original methods are maintained (including attributes), but their bodies are replaced with `throw NotImplementedException()`.  
+- Those should not be referenced in IL.  
+- They are marked with `[ExtensionMarkerName("...")]` referring to the name of the extension marker type corresponding to the extension block for that member.  
 
-#### Implementations
+#### Implementation methods
 
-The method bodies for method/property declarations in an extension declaration in source are emitted 
+The method bodies for method/property declarations in an extension block in source are emitted 
 as static implementation methods in the top-level static class.  
 - An implementation method has the same name as the original method.  
-- It has type parameters derived from the extension declaration prepended to the type parameters of the original method (including attributes).  
+- It has type parameters derived from the extension block prepended to the type parameters of the original method (including attributes).  
 - It has the same accessibility and attributes as the original method.  
 - If it implements a static method, it has the same parameters and return type. 
 - It if implements an instance method, it has a prepended parameter to the signature of the original method. 
-  This parameter's attributes, refness, type, and name are derived from the receiver parameter declared in the relevant extension declaration.
-- The parameters in implementation methods refer to type parameters owned by implementation method, instead of those of an extension declaration.  
+  This parameter's attributes, refness, type, and name are derived from the extension parameter declared in the relevant extension block.
+- The parameters in implementation methods refer to type parameters owned by implementation method, instead of those of an extension block.  
 - If the original member is an instance ordinary method, the implementation method is marked with an `[Extension]` attribute.
 
-For example:
+#### Example
+
+Note: we use simplified content-based names for the example, for readability.
+Note: since C# cannot represent type parameter re-declaration, the code representing metadata is not valid C# code.
+
+Here's an example illustrating how grouping works, without members:
+
 ```
-static class IEnumerableExtensions
+class E
 {
     extension<T>(IEnumerable<T> source)
+    {
+        ... member in extension<T>(IEnumerable<T> source)
+    }
+
+    extension<U>(ref IEnumerable<U?> p)
+    {
+        ... member in extension<U>(ref IEnumerable<U?> p)
+    }
+
+    extension<T>(IEnumerable<U> source)
+        where T : IEquatable<U>
+    {
+        ... member in extension<T>(IEnumerable<U> source) where T : IEquatable<U>
+    }
+}
+```
+is emitted as
+```
+class E
+{
+    public class <>E__ContentName_For_IEnumerable_T<T0>
+    {
+        private static class <>E__ContentName1 // note: re-declares type parameters
```

Comment:
>// note: re-declares type parameters [](http://example.com/codeflow?start=47&length=36)

Not sure how to interpret the comment, I do not see the redeclaration.

Reply:
I don't want to show IL syntax, so I'm using C# syntax and indicating that this nested type re-declared type parameters. In this example, the marker method refers to `T` which results from that re-declaration.
I have a comment above: "since C# cannot represent type parameter re-declaration, the code representing metadata is not valid C# code."